// @framer/steveruizok.router/code/RouterOverrides.tsx
import { Data, Override } from 'framer'

let router

export const isNavigation: Override = props => {
	return {
		onMount(myRouter: any) {
			router = myRouter
			// console.log('Router mounted', router)
		},
		onTransitionStart(state: any) {
			// console.log('Transition started', state)
		},
		onTransitionEnd(state: any) {
			// console.log('Transition ended', state)
		},
	}
}

export const isBackButton: Override = props => {
	return {
		onTap() {
			router.showPrevious()
		},
	}
}

export const isTab1Button: Override = props => {
	return {
		onTap() {
			router.showNext(0)
		},
	}
}

export const isArticle1Button: Override = props => {
	return {
		onTap() {
			router.showNext(4)
		},
	}
}

export const isTab2Button: Override = props => {
	return {
		onTap() {
			router.showNext(1)
		},
	}
}

export const isArticle2Button: Override = props => {
	return {
		onTap() {
			router.showNext(5)
		},
	}
}

export const isTab3Button: Override = props => {
	return {
		onTap() {
			router.showNext(2)
		},
	}
}

export const isArticle3Button: Override = props => {
	return {
		onTap() {
			router.showNext(6)
		},
	}
}
export const isTab4Button: Override = props => {
	return {
		onTap() {
			router.showNext(3)
		},
	}
}

export const isArticle4Button: Override = props => {
	return {
		onTap() {
			router.showNext(7)
		},
	}
}
